### PR TITLE
Implement light client friendliness changes

### PR DESCRIPTION
**Pull request by hwwhww**
### What was wrong?

Implement https://github.com/ethereum/eth2.0-specs/pull/476
Rationale: https://github.com/ethereum/eth2.0-specs/issues/459

Note that it's this version of the spec: https://github.com/ethereum/eth2.0-specs/blob/85d39af1cad0e4869944be8bdcbfb75b7c3c64f6/specs/core/0_beacon-chain.md and it doesn't reflect #235 yet.

### How was it fixed?
- Remove
    - `ValidatorRegistryDeltaBlock`
    - `ValidatorRegistryDeltaFlag`
    - `BeaconState.validator_registry_delta_chain_tip`
- Rename
    - `BeaconState.previous_epoch_randao_mix` -> `BeaconState.previous_epoch_seed`
    - `BeaconState.current_epoch_randao_mix ` -> `BeaconState.current_epoch_seed`
- Add
    - `BeaconState.latest_index_roots`
    - Helpers:
        - `get_active_index_root`
        - `get_entry_exit_effect_slot`
        - `generate_seed`: note that it was added by other spec PR, but added the "slot" version here since it's helpful to refactor the code. Will have to change to "epoch" version for #235 in other PR.
- Update
    - `eth2/beacon/on_startup.py`: update the state fields.
    - `validator_status_helpers.py`: remove `validator_registry_delta_chain_tip` updating.

#### Cute Animal Picture

larry_the_cat_standing_on_gravel_and_verociously_attacking_a_wool_string_in_auderghem _belgium_ dscf2327
[source](https://commons.wikimedia.org/wiki/File:Larry_the_cat_standing_on_gravel_and_verociously_attacking_a_wool_string_in_Auderghem,_Belgium_(DSCF2327).jpg)

### CODE REVIEW

File: tests/eth2/beacon/conftest.py [hunk 1]
```diff
@@ -341,13 +341,15 @@ def filled_beacon_state(genesis_slot,
                         genesis_start_shard,
```
Comment (ChihChengLiang) [inline tests/eth2/beacon/conftest.py]: ☝️ sample_validator_registry_delta_block_params should be nuked

File: eth2/beacon/on_startup.py [hunk 2]
```diff
@@ -154,8 +154,9 @@ def get_initial_beacon_state(*,
             state = activate_validator(
                 state,
                 validator_index,
-                genesis=True,
+                is_genesis=True,
```
Comment (pipermerriam) [inline eth2/beacon/on_startup.py]: :+1:

File: eth2/beacon/state_machines/forks/serenity/epoch_processing.py [hunk 3]
```diff
@@ -88,13 +127,21 @@ def process_validator_registry(state: BeaconState,
             state = state.copy(
                 current_epoch_calculation_slot=state.slot,
             )
+
+            # `helpers.generate_seed` path for mocking tests
```
Comment (pipermerriam) [inline eth2/beacon/state_machines/forks/serenity/epoch_processing.py]: I *think* know what this comment is indicating but could we maybe make it a bit more explicit.  Here's what I think it is trying to say.

> The `helpers.generate_seed` function is only present to provide an entry point for mocking this out in tests.

File: eth2/beacon/state_machines/forks/serenity/epoch_processing.py [hunk 4]
```diff
@@ -48,13 +52,42 @@ def update_validator_registry(state: BeaconState) -> BeaconState:
...
+def _update_latest_index_roots(state: BeaconState,
+                               config: BeaconConfig) -> BeaconState:
+    """
+    Return the BeaconState with updated `latest_index_roots`.
+    """
+    next_epoch = state.next_epoch(config.EPOCH_LENGTH)
+
+    # TODO: chanege to hash_tree_root
+    index_root = hash_eth2(
+        bytes(
```
Comment (pipermerriam) [inline eth2/beacon/state_machines/forks/serenity/epoch_processing.py]: Is it guaranteed that `get_active_validator_indices()` will only return values in the range `0-255`?  A glance at the implementation doesn't appear it has that guarantee.  If not, then this call will fail if any of the indices are outside of this range.

File: tests/eth2/beacon/state_machines/forks/test_serenity_epoch_processing.py [hunk 5]
```diff
@@ -94,40 +99,76 @@ def test_check_if_update_validator_registry(genesis_state,
...
+                                   epoch_length,
+                                   latest_index_roots_length):
+    state = genesis_state.copy(
+        slot=state_slot,
+    )
+
+    result_state = _update_latest_index_roots(state, config)
+
+    index_root = hash_eth2(
+        bytes(
```
Comment (pipermerriam) [inline tests/eth2/beacon/state_machines/forks/test_serenity_epoch_processing.py]: Same here WRT to having any indices that are outside of `0-255`

File: eth2/beacon/state_machines/forks/serenity/epoch_processing.py (hunk 4 above)
Comment (hwwhww) [reply eth2/beacon/state_machines/forks/serenity/epoch_processing.py]: ohhh good point!!!!
(Note that it's a workaround before we add tree hashing.) I'm changing it to:
```python
    # TODO: chanege to hash_tree_root
    index_root = hash_eth2(
        b''.join(
            [
                index.to_bytes(32, 'big')
                for index in get_active_validator_indices(
                    state.validator_registry,
                    # TODO: change to `per-epoch` version
                    state.slot,
                )
            ]
        )
    )
```

(p.s. The upper-bound of the size of `ValidatorIndex` is uint64.)

**Pull request merged**